### PR TITLE
283 Change CurrencyId enum

### PR DESCRIPTION
**Pull request by b-yap**
the command i use to expand the spacewalk project:
`spacewalk/primitives %  cargo rustc --profile=check -- -Zunpretty=expanded`

From branch `dev/milestone-3` of spacewalk:
```rust
pub enum CurrencyId { Native, XCM(ForeignCurrencyId), Stellar(Asset), }
#[automatically_derived]
impl ::codec::Encode for CurrencyId {
    fn encode_to<__CodecOutputEdqy: ::codec::Output +
        ?::core::marker::Sized>(&self,
        __codec_dest_edqy: &mut __CodecOutputEdqy) {
        match *self {
            CurrencyId::Native => {
                __codec_dest_edqy.push_byte(0usize as
                        ::core::primitive::u8);
            }
            CurrencyId::XCM(ref aa) => {
                __codec_dest_edqy.push_byte(1usize as
                        ::core::primitive::u8);
                ::codec::Encode::encode_to(aa, __codec_dest_edqy);
            }
            CurrencyId::Stellar(ref aa) => {
                __codec_dest_edqy.push_byte(2usize as
                        ::core::primitive::u8);
                ::codec::Encode::encode_to(aa, __codec_dest_edqy);
            }
            _ => (),
        }
    }
}

#[automatically_derived]
impl ::codec::Decode for CurrencyId {
    fn decode<__CodecInputEdqy: ::codec::Input>(__codec_input_edqy:
            &mut __CodecInputEdqy)
        -> ::core::result::Result<Self, ::codec::Error> {
        match __codec_input_edqy.read_byte().map_err(|e|
                        e.chain("Could not decode `CurrencyId`, failed to read variant byte"))?
            {
            __codec_x_edqy if
                __codec_x_edqy == 0usize as ::core::primitive::u8 => {
                ::core::result::Result::Ok(CurrencyId::Native)
            }
            __codec_x_edqy if
                __codec_x_edqy == 1usize as ::core::primitive::u8 => {
                ::core::result::Result::Ok(CurrencyId::XCM({
                            let __codec_res_edqy =
                                <ForeignCurrencyId as
                                        ::codec::Decode>::decode(__codec_input_edqy);
                            match __codec_res_edqy {
                                ::core::result::Result::Err(e) =>
                                    return ::core::result::Result::Err(e.chain("Could not decode `CurrencyId::XCM.0`")),
                                ::core::result::Result::Ok(__codec_res_edqy) =>
                                    __codec_res_edqy,
                            }
                        }))
            }
            __codec_x_edqy if
                __codec_x_edqy == 2usize as ::core::primitive::u8 => {
                ::core::result::Result::Ok(CurrencyId::Stellar({
                            let __codec_res_edqy =
                                <Asset as ::codec::Decode>::decode(__codec_input_edqy);
                            match __codec_res_edqy {
                                ::core::result::Result::Err(e) =>
                                    return ::core::result::Result::Err(e.chain("Could not decode `CurrencyId::Stellar.0`")),
                                ::core::result::Result::Ok(__codec_res_edqy) =>
                                    __codec_res_edqy,
                            }
                        }))
            }
            _ =>
                ::core::result::Result::Err(<_ as
                            ::core::convert::Into<_>>::into("Could not decode `CurrencyId`, variant doesn't exist")),
        }
    }
}
#[automatically_derived]
impl ::core::marker::StructuralEq for CurrencyId { }
#[automatically_derived]
impl ::core::cmp::Eq for CurrencyId {
    #[inline]
    #[doc(hidden)]
    #[no_coverage]
    fn assert_receiver_is_total_eq(&self) -> () {
        let _: ::core::cmp::AssertParamIsEq<ForeignCurrencyId>;
        let _: ::core::cmp::AssertParamIsEq<Asset>;
    }
}
#[automatically_derived]
impl ::core::hash::Hash for CurrencyId {
    fn hash<__H: ::core::hash::Hasher>(&self, state: &mut __H) -> () {
        let __self_tag = ::core::intrinsics::discriminant_value(self);
        ::core::hash::Hash::hash(&__self_tag, state);
        match self {
            CurrencyId::XCM(__self_0) =>
                ::core::hash::Hash::hash(__self_0, state),
            CurrencyId::Stellar(__self_0) =>
                ::core::hash::Hash::hash(__self_0, state),
            _ => {}
        }
    }
}
#[automatically_derived]
impl ::core::marker::StructuralPartialEq for CurrencyId { }
#[automatically_derived]
impl ::core::cmp::PartialEq for CurrencyId {
    #[inline]
    fn eq(&self, other: &CurrencyId) -> bool {
        let __self_tag = ::core::intrinsics::discriminant_value(self);
        let __arg1_tag = ::core::intrinsics::discriminant_value(other);
        __self_tag == __arg1_tag &&
            match (self, other) {
                (CurrencyId::XCM(__self_0), CurrencyId::XCM(__arg1_0)) =>
                    *__self_0 == *__arg1_0,
                (CurrencyId::Stellar(__self_0), CurrencyId::Stellar(__arg1_0))
                    => *__self_0 == *__arg1_0,
                _ => true,
            }
    }
}
#[automatically_derived]
impl ::core::marker::Copy for CurrencyId { }
#[automatically_derived]
impl ::core::clone::Clone for CurrencyId {
    #[inline]
    fn clone(&self) -> CurrencyId {
        let _: ::core::clone::AssertParamIsClone<ForeignCurrencyId>;
        let _: ::core::clone::AssertParamIsClone<Asset>;
        *self
    }
}
#[automatically_derived]
impl ::core::cmp::PartialOrd for CurrencyId {
    #[inline]
    fn partial_cmp(&self, other: &CurrencyId)
        -> ::core::option::Option<::core::cmp::Ordering> {
        let __self_tag = ::core::intrinsics::discriminant_value(self);
        let __arg1_tag = ::core::intrinsics::discriminant_value(other);
        match ::core::cmp::PartialOrd::partial_cmp(&__self_tag, &__arg1_tag) {
            ::core::option::Option::Some(::core::cmp::Ordering::Equal) =>
                match (self, other) {
                    (CurrencyId::XCM(__self_0), CurrencyId::XCM(__arg1_0)) =>
                        ::core::cmp::PartialOrd::partial_cmp(__self_0, __arg1_0),
                    (CurrencyId::Stellar(__self_0),
                        CurrencyId::Stellar(__arg1_0)) =>
                        ::core::cmp::PartialOrd::partial_cmp(__self_0, __arg1_0),
                    _ =>
                        ::core::option::Option::Some(::core::cmp::Ordering::Equal),
                },
            cmp => cmp,
        }
    }
}
#[automatically_derived]
impl ::core::cmp::Ord for CurrencyId {
    #[inline]
    fn cmp(&self, other: &CurrencyId) -> ::core::cmp::Ordering {
        let __self_tag = ::core::intrinsics::discriminant_value(self);
        let __arg1_tag = ::core::intrinsics::discriminant_value(other);
        match ::core::cmp::Ord::cmp(&__self_tag, &__arg1_tag) {
            ::core::cmp::Ordering::Equal =>
                match (self, other) {
                    (CurrencyId::XCM(__self_0), CurrencyId::XCM(__arg1_0)) =>
                        ::core::cmp::Ord::cmp(__self_0, __arg1_0),
                    (CurrencyId::Stellar(__self_0),
                        CurrencyId::Stellar(__arg1_0)) =>
                        ::core::cmp::Ord::cmp(__self_0, __arg1_0),
                    _ => ::core::cmp::Ordering::Equal,
                },
            cmp => cmp,
        }
    }
}

impl CurrencyId {
    pub const StellarNative: CurrencyId = Self::Stellar(Asset::StellarNative);

    #[allow(non_snake_case)]
    pub const fn AlphaNum4(code: Bytes4, issuer: AssetIssuer) -> Self {
        Self::Stellar(Asset::AlphaNum4 { code, issuer })
    }

    #[allow(non_snake_case)]
    pub const fn AlphaNum12(code: Bytes12, issuer: AssetIssuer) -> Self {
        Self::Stellar(Asset::AlphaNum12 { code, issuer })
    }
}

repr(u8)]
pub enum ForeignCurrencyId {
    KSM = 0,
    KAR = 1,
    AUSD = 2,
    BNC = 3,
    VsKSM = 4,
    HKO = 5,
    MOVR = 6,
    SDN = 7,
    KINT = 8,
    KBTC = 9,
    GENS = 10,
    XOR = 11,
    TEER = 12,
    KILT = 13,
    PHA = 14,
    ZTG = 15,
    USDT = 16,
    DOT =
    // added lastly
    20,
}
#[automatically_derived]
impl ::codec::Encode for ForeignCurrencyId {
    fn encode_to<__CodecOutputEdqy: ::codec::Output +
        ?::core::marker::Sized>(&self,
        __codec_dest_edqy: &mut __CodecOutputEdqy) {
        match *self {
            ForeignCurrencyId::KSM => {
                __codec_dest_edqy.push_byte(0 as ::core::primitive::u8);
            }
            ForeignCurrencyId::KAR => {
                __codec_dest_edqy.push_byte(1 as ::core::primitive::u8);
            }
            ForeignCurrencyId::AUSD => {
                __codec_dest_edqy.push_byte(2 as ::core::primitive::u8);
            }
            ForeignCurrencyId::BNC => {
                __codec_dest_edqy.push_byte(3 as ::core::primitive::u8);
            }
            ForeignCurrencyId::VsKSM => {
                __codec_dest_edqy.push_byte(4 as ::core::primitive::u8);
            }
            ForeignCurrencyId::HKO => {
                __codec_dest_edqy.push_byte(5 as ::core::primitive::u8);
            }
            ForeignCurrencyId::MOVR => {
                __codec_dest_edqy.push_byte(6 as ::core::primitive::u8);
            }
            ForeignCurrencyId::SDN => {
                __codec_dest_edqy.push_byte(7 as ::core::primitive::u8);
            }
            ForeignCurrencyId::KINT => {
                __codec_dest_edqy.push_byte(8 as ::core::primitive::u8);
            }
            ForeignCurrencyId::KBTC => {
                __codec_dest_edqy.push_byte(9 as ::core::primitive::u8);
            }
            ForeignCurrencyId::GENS => {
                __codec_dest_edqy.push_byte(10 as ::core::primitive::u8);
            }
            ForeignCurrencyId::XOR => {
                __codec_dest_edqy.push_byte(11 as ::core::primitive::u8);
            }
            ForeignCurrencyId::TEER => {
                __codec_dest_edqy.push_byte(12 as ::core::primitive::u8);
            }
            ForeignCurrencyId::KILT => {
                __codec_dest_edqy.push_byte(13 as ::core::primitive::u8);
            }
            ForeignCurrencyId::PHA => {
                __codec_dest_edqy.push_byte(14 as ::core::primitive::u8);
            }
            ForeignCurrencyId::ZTG => {
                __codec_dest_edqy.push_byte(15 as ::core::primitive::u8);
            }
            ForeignCurrencyId::USDT => {
                __codec_dest_edqy.push_byte(16 as ::core::primitive::u8);
            }
            ForeignCurrencyId::DOT => {
                __codec_dest_edqy.push_byte(20 as ::core::primitive::u8);
            }
            _ => (),
        }
    }
}

#[automatically_derived]
impl ::codec::Decode for ForeignCurrencyId {
    fn decode<__CodecInputEdqy: ::codec::Input>(__codec_input_edqy:
            &mut __CodecInputEdqy)
        -> ::core::result::Result<Self, ::codec::Error> {
        match __codec_input_edqy.read_byte().map_err(|e|
                        e.chain("Could not decode `ForeignCurrencyId`, failed to read variant byte"))?
            {
            __codec_x_edqy if
                __codec_x_edqy == 0 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::KSM)
            }
            __codec_x_edqy if
                __codec_x_edqy == 1 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::KAR)
            }
            __codec_x_edqy if
                __codec_x_edqy == 2 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::AUSD)
            }
            __codec_x_edqy if
                __codec_x_edqy == 3 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::BNC)
            }
            __codec_x_edqy if
                __codec_x_edqy == 4 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::VsKSM)
            }
            __codec_x_edqy if
                __codec_x_edqy == 5 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::HKO)
            }
            __codec_x_edqy if
                __codec_x_edqy == 6 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::MOVR)
            }
            __codec_x_edqy if
                __codec_x_edqy == 7 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::SDN)
            }
            __codec_x_edqy if
                __codec_x_edqy == 8 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::KINT)
            }
            __codec_x_edqy if
                __codec_x_edqy == 9 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::KBTC)
            }
            __codec_x_edqy if
                __codec_x_edqy == 10 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::GENS)
            }
            __codec_x_edqy if
                __codec_x_edqy == 11 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::XOR)
            }
            __codec_x_edqy if
                __codec_x_edqy == 12 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::TEER)
            }
            __codec_x_edqy if
                __codec_x_edqy == 13 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::KILT)
            }
            __codec_x_edqy if
                __codec_x_edqy == 14 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::PHA)
            }
            __codec_x_edqy if
                __codec_x_edqy == 15 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::ZTG)
            }
            __codec_x_edqy if
                __codec_x_edqy == 16 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::USDT)
            }
            __codec_x_edqy if
                __codec_x_edqy == 20 as ::core::primitive::u8 => {
                ::core::result::Result::Ok(ForeignCurrencyId::DOT)
            }
            _ =>
                ::core::result::Result::Err(<_ as
                            ::core::convert::Into<_>>::into("Could not decode `ForeignCurrencyId`, variant doesn't exist")),
        }
    }
}

#[automatically_derived]
impl ::core::marker::StructuralEq for ForeignCurrencyId { }
#[automatically_derived]
impl ::core::cmp::Eq for ForeignCurrencyId {
    #[inline]
    #[doc(hidden)]
    #[no_coverage]
    fn assert_receiver_is_total_eq(&self) -> () {}
}
#[automatically_derived]
impl ::core::hash::Hash for ForeignCurrencyId {
    fn hash<__H: ::core::hash::Hasher>(&self, state: &mut __H) -> () {
        let __self_tag = ::core::intrinsics::discriminant_value(self);
        ::core::hash::Hash::hash(&__self_tag, state)
    }
}
#[automatically_derived]
impl ::core::marker::StructuralPartialEq for ForeignCurrencyId { }
#[automatically_derived]
impl ::core::cmp::PartialEq for ForeignCurrencyId {
    #[inline]
    fn eq(&self, other: &ForeignCurrencyId) -> bool {
        let __self_tag = ::core::intrinsics::discriminant_value(self);
        let __arg1_tag = ::core::intrinsics::discriminant_value(other);
        __self_tag == __arg1_tag
    }
}
#[automatically_derived]
impl ::core::marker::Copy for ForeignCurrencyId { }
#[automatically_derived]
impl ::core::clone::Clone for ForeignCurrencyId {
    #[inline]
    fn clone(&self) -> ForeignCurrencyId { *self }
}
#[automatically_derived]
impl ::core::cmp::PartialOrd for ForeignCurrencyId {
    #[inline]
    fn partial_cmp(&self, other: &ForeignCurrencyId)
        -> ::core::option::Option<::core::cmp::Ordering> {
        let __self_tag = ::core::intrinsics::discriminant_value(self);
        let __arg1_tag = ::core::intrinsics::discriminant_value(other);
        ::core::cmp::PartialOrd::partial_cmp(&__self_tag, &__arg1_tag)
    }
}
#[automatically_derived]
impl ::core::cmp::Ord for ForeignCurrencyId {
    #[inline]
    fn cmp(&self, other: &ForeignCurrencyId) -> ::core::cmp::Ordering {
        let __self_tag = ::core::intrinsics::discriminant_value(self);
        let __arg1_tag = ::core::intrinsics::discriminant_value(other);
        ::core::cmp::Ord::cmp(&__self_tag, &__arg1_tag)
    }
}
pub const KSM: ForeignCurrencyId = ForeignCurrencyId::KSM;
pub const KAR: ForeignCurrencyId = ForeignCurrencyId::KAR;
pub const AUSD: ForeignCurrencyId = ForeignCurrencyId::AUSD;
pub const BNC: ForeignCurrencyId = ForeignCurrencyId::BNC;
pub const VsKSM: ForeignCurrencyId = ForeignCurrencyId::VsKSM;
pub const HKO: ForeignCurrencyId = ForeignCurrencyId::HKO;
pub const MOVR: ForeignCurrencyId = ForeignCurrencyId::MOVR;
pub const SDN: ForeignCurrencyId = ForeignCurrencyId::SDN;
pub const KINT: ForeignCurrencyId = ForeignCurrencyId::KINT;
pub const KBTC: ForeignCurrencyId = ForeignCurrencyId::KBTC;
pub const GENS: ForeignCurrencyId = ForeignCurrencyId::GENS;
pub const XOR: ForeignCurrencyId = ForeignCurrencyId::XOR;
pub const TEER: ForeignCurrencyId = ForeignCurrencyId::TEER;
pub const KILT: ForeignCurrencyId = ForeignCurrencyId::KILT;
pub const PHA: ForeignCurrencyId = ForeignCurrencyId::PHA;
pub const ZTG: ForeignCurrencyId = ForeignCurrencyId::ZTG;
pub const USDT: ForeignCurrencyId = ForeignCurrencyId::USDT;
pub const DOT: ForeignCurrencyId = ForeignCurrencyId::DOT;
impl TryFrom<u8> for ForeignCurrencyId {
    type Error = ();
    fn try_from(v: u8) -> Result<Self, Self::Error> {
        match v {
            0 => Ok(ForeignCurrencyId::KSM),
            1 => Ok(ForeignCurrencyId::KAR),
            2 => Ok(ForeignCurrencyId::AUSD),
            3 => Ok(ForeignCurrencyId::BNC),
            4 => Ok(ForeignCurrencyId::VsKSM),
            5 => Ok(ForeignCurrencyId::HKO),
            6 => Ok(ForeignCurrencyId::MOVR),
            7 => Ok(ForeignCurrencyId::SDN),
            8 => Ok(ForeignCurrencyId::KINT),
            9 => Ok(ForeignCurrencyId::KBTC),
            10 => Ok(ForeignCurrencyId::GENS),
            11 => Ok(ForeignCurrencyId::XOR),
            12 => Ok(ForeignCurrencyId::TEER),
            13 => Ok(ForeignCurrencyId::KILT),
            14 => Ok(ForeignCurrencyId::PHA),
            15 => Ok(ForeignCurrencyId::ZTG),
            16 => Ok(ForeignCurrencyId::USDT),
            20 => Ok(ForeignCurrencyId::DOT),
            _ => Err(()),
        }
    }
}
impl TryFrom<u64> for ForeignCurrencyId {
    type Error = ();
    fn try_from(v: u64) -> Result<Self, Self::Error> {
        let v_u8: u8 = u8::try_from(v).map_err(|_| ())?;
        v_u8.try_into()
    }
}
impl Into<u8> for ForeignCurrencyId {
    fn into(self) -> u8 {
        match self {
            ForeignCurrencyId::KSM => (0),
            ForeignCurrencyId::KAR => (1),
            ForeignCurrencyId::AUSD => (2),
            ForeignCurrencyId::BNC => (3),
            ForeignCurrencyId::VsKSM => (4),
            ForeignCurrencyId::HKO => (5),
            ForeignCurrencyId::MOVR => (6),
            ForeignCurrencyId::SDN => (7),
            ForeignCurrencyId::KINT => (8),
            ForeignCurrencyId::KBTC => (9),
            ForeignCurrencyId::GENS => (10),
            ForeignCurrencyId::XOR => (11),
            ForeignCurrencyId::TEER => (12),
            ForeignCurrencyId::KILT => (13),
            ForeignCurrencyId::PHA => (14),
            ForeignCurrencyId::ZTG => (15),
            ForeignCurrencyId::USDT => (16),
            ForeignCurrencyId::DOT => (20),
        }
    }
}
impl ForeignCurrencyId {
    pub fn get_info() -> Vec<(&'static str, u32)> { (/*ERROR*/) }
    pub const fn one(&self) -> Balance { 10u128.pow(self.decimals() as u32) }
    const fn decimals(&self) -> u8 {
        match self {
            ForeignCurrencyId::KSM => 10,
            ForeignCurrencyId::KAR => 10,
            ForeignCurrencyId::AUSD => 10,
            ForeignCurrencyId::BNC => 10,
            ForeignCurrencyId::VsKSM => 10,
            ForeignCurrencyId::HKO => 10,
            ForeignCurrencyId::MOVR => 10,
            ForeignCurrencyId::SDN => 10,
            ForeignCurrencyId::KINT => 10,
            ForeignCurrencyId::KBTC => 10,
            ForeignCurrencyId::GENS => 10,
            ForeignCurrencyId::XOR => 10,
            ForeignCurrencyId::TEER => 10,
            ForeignCurrencyId::KILT => 10,
            ForeignCurrencyId::PHA => 10,
            ForeignCurrencyId::ZTG => 10,
            ForeignCurrencyId::USDT => 10,
            ForeignCurrencyId::DOT => 10,
        }
    }
}
impl CurrencyInfo for ForeignCurrencyId {
    fn name(&self) -> &str {
        match self {
            ForeignCurrencyId::KSM => "Kusama",
            ForeignCurrencyId::KAR => "Karura",
            ForeignCurrencyId::AUSD => "Acala Karura",
            ForeignCurrencyId::BNC => "Bifrost",
            ForeignCurrencyId::VsKSM => "Kusama Bifrost",
            ForeignCurrencyId::HKO => "Heiko",
            ForeignCurrencyId::MOVR => "Moonriver",
            ForeignCurrencyId::SDN => "Shiden",
            ForeignCurrencyId::KINT => "Kintsugi",
            ForeignCurrencyId::KBTC => "Kintsugi BTC",
            ForeignCurrencyId::GENS => "Genshiro",
            ForeignCurrencyId::XOR => "Sora",
            ForeignCurrencyId::TEER => "Integritee",
            ForeignCurrencyId::KILT => "Kilt",
            ForeignCurrencyId::PHA => "Phala",
            ForeignCurrencyId::ZTG => "Zeitgeist",
            ForeignCurrencyId::USDT => "Statemine",
            ForeignCurrencyId::DOT => "Polkadot",
        }
    }
    fn symbol(&self) -> &str {
        match self {
            ForeignCurrencyId::KSM => "KSM",
            ForeignCurrencyId::KAR => "KAR",
            ForeignCurrencyId::AUSD => "AUSD",
            ForeignCurrencyId::BNC => "BNC",
            ForeignCurrencyId::VsKSM => "VsKSM",
            ForeignCurrencyId::HKO => "HKO",
            ForeignCurrencyId::MOVR => "MOVR",
            ForeignCurrencyId::SDN => "SDN",
            ForeignCurrencyId::KINT => "KINT",
            ForeignCurrencyId::KBTC => "KBTC",
            ForeignCurrencyId::GENS => "GENS",
            ForeignCurrencyId::XOR => "XOR",
            ForeignCurrencyId::TEER => "TEER",
            ForeignCurrencyId::KILT => "KILT",
            ForeignCurrencyId::PHA => "PHA",
            ForeignCurrencyId::ZTG => "ZTG",
            ForeignCurrencyId::USDT => "USDT",
            ForeignCurrencyId::DOT => "DOT",
        }
    }
    fn decimals(&self) -> u8 { self.decimals() }
}
impl TryFrom<Vec<u8>> for ForeignCurrencyId {
    type Error = ();
    fn try_from(v: Vec<u8>) -> Result<ForeignCurrencyId, ()> {
        match v.as_slice() {
            _ => Ok(ForeignCurrencyId::KSM),
            _ => Ok(ForeignCurrencyId::KAR),
            _ => Ok(ForeignCurrencyId::AUSD),
            _ => Ok(ForeignCurrencyId::BNC),
            _ => Ok(ForeignCurrencyId::VsKSM),
            _ => Ok(ForeignCurrencyId::HKO),
            _ => Ok(ForeignCurrencyId::MOVR),
            _ => Ok(ForeignCurrencyId::SDN),
            _ => Ok(ForeignCurrencyId::KINT),
            _ => Ok(ForeignCurrencyId::KBTC),
            _ => Ok(ForeignCurrencyId::GENS),
            _ => Ok(ForeignCurrencyId::XOR),
            _ => Ok(ForeignCurrencyId::TEER),
            _ => Ok(ForeignCurrencyId::KILT),
            _ => Ok(ForeignCurrencyId::PHA),
            _ => Ok(ForeignCurrencyId::ZTG),
            _ => Ok(ForeignCurrencyId::USDT),
            _ => Ok(ForeignCurrencyId::DOT),
            _ => Err(()),
        }
    }
}
```

### CODE REVIEW

**Comment by b-yap** [comment]
From branch `main` of pendulum -> runtime/pendulum/currency.rs
```rust
pub enum CurrencyId { Native, XCM(ForeignCurrencyId), }
#[automatically_derived]
impl ::core::marker::StructuralEq for CurrencyId { }
#[automatically_derived]
impl ::core::cmp::Eq for CurrencyId {
    #[inline]
    #[doc(hidden)]
    #[no_coverage]
    fn assert_receiver_is_total_eq(&self) -> () {
        let _: ::core::cmp::AssertParamIsEq<ForeignCurrencyId>;
    }
}
#[automatically_derived]
impl ::core::marker::StructuralPartialEq for CurrencyId { }
#[automatically_derived]
impl ::core::cmp::PartialEq for CurrencyId {
    #[inline]
    fn eq(&self, other: &CurrencyId) -> bool {
        let __self_tag = ::core::intrinsics::discriminant_value(self);
        let __arg1_tag = ::core::intrinsics::discriminant_value(other);
        __self_tag == __arg1_tag &&
            match (self, other) {
                (CurrencyId::XCM(__self_0), CurrencyId::XCM(__arg1_0)) =>
                    *__self_0 == *__arg1_0,
                _ => true,
            }
    }
}
#[automatically_derived]
impl ::core::marker::Copy for CurrencyId { }
#[automatically_derived]
impl ::core::clone::Clone for CurrencyId {
    #[inline]
    fn clone(&self) -> CurrencyId {
        let _: ::core::clone::AssertParamIsClone<ForeignCurrencyId>;
        *self
    }
}
#[automatically_derived]
impl ::core::cmp::PartialOrd for CurrencyId {
    #[inline]
    fn partial_cmp(&self, other: &CurrencyId)
        -> ::core::option::Option<::core::cmp::Ordering> {
        let __self_tag = ::core::intrinsics::discriminant_value(self);
        let __arg1_tag = ::core::intrinsics::discriminant_value(other);
        match ::core::cmp::PartialOrd::partial_cmp(&__self_tag, &__arg1_tag) {
            ::core::option::Option::Some(::core::cmp::Ordering::Equal) =>
                match (self, other) {
                    (CurrencyId::XCM(__self_0), CurrencyId::XCM(__arg1_0)) =>
                        ::core::cmp::PartialOrd::partial_cmp(__self_0, __arg1_0),
                    _ =>
                        ::core::option::Option::Some(::core::cmp::Ordering::Equal),
                },
            cmp => cmp,
        }
    }
}
#[automatically_derived]
impl ::core::cmp::Ord for CurrencyId {
    #[inline]
    fn cmp(&self, other: &CurrencyId) -> ::core::cmp::Ordering {
        let __self_tag = ::core::intrinsics::discriminant_value(self);
        let __arg1_tag = ::core::intrinsics::discriminant_value(other);
        match ::core::cmp::Ord::cmp(&__self_tag, &__arg1_tag) {
            ::core::cmp::Ordering::Equal =>
                match (self, other) {
                    (CurrencyId::XCM(__self_0), CurrencyId::XCM(__arg1_0)) =>
                        ::core::cmp::Ord::cmp(__self_0, __arg1_0),
                    _ => ::core::cmp::Ordering::Equal,
                },
            cmp => cmp,
        }
    }
}

impl Default for CurrencyId {
    fn default() -> Self { CurrencyId::Native }
}

#[automatically_derived]
impl ::core::marker::StructuralEq for ForeignCurrencyId { }
#[automatically_derived]
impl ::core::cmp::Eq for ForeignCurrencyId {
    #[inline]
    #[doc(hidden)]
    #[no_coverage]
    fn assert_receiver_is_total_eq(&self) -> () {}
}
#[automatically_derived]
impl ::core::marker::StructuralPartialEq for ForeignCurrencyId { }
#[automatically_derived]
impl ::core::cmp::PartialEq for ForeignCurrencyId {
    #[inline]
    fn eq(&self, other: &ForeignCurrencyId) -> bool {
        let __self_tag = ::core::intrinsics::discriminant_value(self);
        let __arg1_tag = ::core::intrinsics::discriminant_value(other);
        __self_tag == __arg1_tag
    }
}
#[automatically_derived]
impl ::core::marker::Copy for ForeignCurrencyId { }
#[automatically_derived]
impl ::core::clone::Clone for ForeignCurrencyId {
    #[inline]
    fn clone(&self) -> ForeignCurrencyId { *self }
}
#[automatically_derived]
impl ::core::cmp::PartialOrd for ForeignCurrencyId {
    #[inline]
    fn partial_cmp(&self, other: &ForeignCurrencyId)
        -> ::core::option::Option<::core::cmp::Ordering> {
        let __self_tag = ::core::intrinsics::discriminant_value(self);
        let __arg1_tag = ::core::intrinsics::discriminant_value(other);
        ::core::cmp::PartialOrd::partial_cmp(&__self_tag, &__arg1_tag)
    }
}
#[automatically_derived]
impl ::core::cmp::Ord for ForeignCurrencyId {
    #[inline]
    fn cmp(&self, other: &ForeignCurrencyId) -> ::core::cmp::Ordering {
        let __self_tag = ::core::intrinsics::discriminant_value(self);
        let __arg1_tag = ::core::intrinsics::discriminant_value(other);
        ::core::cmp::Ord::cmp(&__self_tag, &__arg1_tag)
    }
}
```

File: primitives/src/lib.rs [hunk 1]
```diff
@@ -539,8 +539,8 @@ impl Asset {
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
 #[cfg_attr(feature = "std", serde(rename_all = "camelCase"))]
 pub enum CurrencyId {
-	XCM(ForeignCurrencyId),
 	Native,
+	XCM(ForeignCurrencyId),
 	Stellar(Asset),
```
Comment (ebma) [inline primitives/src/lib.rs:544]: Can you now add explicit numbers here? ie
```
Native = 0,
XCM(ForeignCurrencyId) = 1,
Stellar(Asset) = 2,
```

Let's also add numbers to the `Asset` enum just to be safe.

**Review by ebma — CHANGES_REQUESTED** [review]

**Comment by ebma** [comment]
So if I understand correctly, the important part of the expanded code in pendulum is this:
```
match *self {
            CurrencyId::Native => {
                __codec_dest_edqy.push_byte( ** 0usize ** as
                        ::core::primitive::u8);
            }
            CurrencyId::XCM(ref aa) => {
                __codec_dest_edqy.push_byte( ** 1usize  ** as
                        ::core::primitive::u8);
                ::codec::Encode::encode_to(aa, __codec_dest_edqy);
            }
            _ => (),
        }
```

which tells us that indeed the enum items get assigned a number by their order incrementing the index by one. Thus, by changing the order in the Spacewalk CurrencyId thing (and also optionally assigning fixed indices to them) we should be good to go.

File: primitives/src/lib.rs (hunk 1 above)
Comment (b-yap) [reply primitives/src/lib.rs:544]: ah, can't probably assign a number for non "unit" variant

Comment (b-yap) [reply primitives/src/lib.rs:544]: i can only set "0" for the first ones

**Comment by ebma** [comment]
The integration tests don't work because something is wrong with feeding the price for the XCM asset. You can check this by logging price information (as I did in my last commit, which should be removed again). Because the XCM price is not fed, the parachain status is not set to running and the tests fail.

**Comment by ebma** [comment]
I fixed the issue with the failing integration tests now and this is ready for re-review/merge.

**Review by ebma — APPROVED** [review]

**Pull request merged**